homeassistant/components/stream/core.py

```diff
@@ -356,3 +359,60 @@ async def handle(
     ) -> web.StreamResponse:
         """Handle the stream request."""
         raise NotImplementedError()
+
+
+class KeyFrameConverter:
+    """Generate and hold the keyframe as a jpeg."""
+
+    def __init__(self) -> None:
+        """Initialize."""
+
+        # Keep import here so that we can import stream integration without installing reqs
+        # pylint: disable=import-outside-toplevel
+        from homeassistant.components.camera.img_util import TurboJPEGSingleton
+
+        self.image_requested = False
+        self.packet: Packet = None
+        self._image = b""
+        self._event = asyncio.Event()
+        self.generate_image = self._generate_image
+        self._turbojpeg = TurboJPEGSingleton.instance()
+        if not self._turbojpeg:
+            self.generate_image = lambda: None
+
+    async def get_image(self) -> bytes:
+        """
+        Fetch an image from the Stream and return it as a jpeg in bytes.
+
+        This is called from outside the worker.
+        """
+        self.image_requested = True
+        try:
+            async with async_timeout.timeout(KEYFRAME_TIMEOUT):
+                await self._event.wait()
+        except asyncio.TimeoutError:
+            return b""
+        return self._image
+
+    def check_generate_image(self) -> None:
+        """Generate the image if necessary and signal it is done from the worker."""
+        if self.packet:
+            self.generate_image()
+            self.packet = None
+        if self._image:
+            self.image_requested = False
+            self._event.set()
+            self._event.clear()
+
+    def _generate_image(self) -> None:
+        """Generate the keyframe image in the worker."""
+        # decode packet (try up to 3 times)
+        for _i in range(3):
+            if frames := self.packet.decode():
+                break
+        if frames:
+            image_file = BytesIO()
+            bgr_array = frames[0].to_ndarray(format="bgr24")
+            image_file.write(self._turbojpeg.encode(bgr_array))
+            self._image = image_file.getvalue()
```

Comment:
I'd suggest making this the return value of this function and push the internal state management into the caller, so then its easy to see all event.set/event.clear/event.wait/image interactions closer together. It will also make it more clear that `self._image` may not always be updated, since sometimes this function does not find frames.

Reply:
Sure. This will require another local variable since we need to store the result before we close the image_file, but I agree that the end result is clearer.